chainstate/storage/src/internal/store_tx/mod.rs

```diff
@@ -60,7 +60,15 @@ mod well_known {
 pub struct StoreTxRo<'st, B: storage::Backend>(pub(super) storage::TransactionRo<'st, B, Schema>);
 
 /// Read-write chainstate storage transaction
-pub struct StoreTxRw<'st, B: storage::Backend>(pub(super) storage::TransactionRw<'st, B, Schema>);
+pub struct StoreTxRw<'st, B: storage::Backend> {
+    // The underlying storage transaction
+    db_tx: storage::TransactionRw<'st, B, Schema>,
+
+    // Track if an error was encountered during the execution of the transaction. If so, it will be
+    // recorded here and returned by all subsequent operations up to and including the final
+    // `.commit()` or `.abort()`.
+    error: crate::Result<()>,
+}
```

Comment:
I'm not sure. Bool flags lose information about the specific error emitted. Seems to me the storage user should have the opportunity to inspect or log the error. It could collect a `Vec` of errors but it seems a bit overkill to me.

Reply:
So, I don't think I have a solution to this problem that we discussed in the call that will be nice and super-clean. 

In general, I would love to create a generic solution where some object/type can be wrapped by something generic that creates some kind of "error futures". This way, the behavior would be the same no matter what. But with rust being rust, this will involve lots of code duplication due to wrapping.

So, in my head, I'm imagining something like an enum, call it `ObjectDelayedResult`, with two arms `Object`, being a generic, and `Error`. But then for this to work automatically, the `ObjectDelayedResult` should automatically have the ability to replicate all the functions in `Object` on itself, but this isn't possible in rust without macros. 

One way we can get this to work is by making all calls in `Object` happen only through a closure. While this mathematically works fine, it may introduce ownership issues since taking ownership of the transaction is originally the job of ChainstateRef, and if it doesn't, it'll make the code uglier. But I'm not 100% opposed to this if @iljakuklic wants to give it a shot.